### ISSUE TITLE
Should function-arg-placeholders also influence lambdas

### ISSUE THREAD

**Issue by ClHg2** [clangd/clangd#1861]
Here is an old but similar [issue](https://github.com/clangd/clangd/issues/922). When I disable this option, I expect clangd stop inserting parameter names for lambda calls.

**Comment by ClHg2** [comment]
I even prefer not to add parentheses for lambdas, as it is annoying when passing it as a callable function, for instance when calling `std::sort` which expects a comparator.

**Comment by HighCommander4** [comment]
I agree that not respecting `--function-arg-placholders` for lambdas seems like an oversight.

Regarding parentheses, there are a couple of related bugs on file:
 * https://github.com/llvm/llvm-project/issues/63565 is about adding an option to disable inserting of parentheses in general (not specific to lambdas)
 * https://github.com/clangd/clangd/issues/968 (and related issues https://github.com/clangd/clangd/issues/439 and https://github.com/clangd/clangd/issues/1677) are about heuristics for inserting parentheses or not based on whether the use of the function name can be a call.
   * Perhaps the heuristics could be different for lambdas because they're more likely to be passed as arguments to functions than called? It's hard to say, people also define lambdas for the purpose of calling them (maybe multiple times) locally in the function where they're defined.

**Comment by HighCommander4** [comment]
@Nodli this should be a pretty easy bug to fix, would you like to try?

I'm happy to provide some code pointers / starting points.

**Comment by Nodli** [comment]
From what I can see :

CodeComplete.cpp
L. 649
Completion.Kind == CompletionitemKind::Variable
needs to move up to
L. 597 in the MayHaveArgList boolean.

I will test (and eventually ask for a pull request) if my potato is able to finish compiling clang.

**Comment by Nodli** [comment]
The idea above fails some tests with template variable completion.

@HighCommander4 
What do you think about fixing the problem before summarizeSnippet by detecting lambdas as CompletionItemKind::Function ? Function-like macros are already detected as CompletinItemKind::Function.

Change essentially looks like this :

 ```
--- a/clang-tools-extra/clangd/CodeComplete.cpp
+++ b/clang-tools-extra/clangd/CodeComplete.cpp
@@ -166,8 +166,11 @@ toCompletionItemKind(index::SymbolKind Kind,
 // index::SymbolKind overload where appropriate.
 CompletionItemKind toCompletionItemKind(const CodeCompletionResult &Res,
                                         CodeCompletionContext::Kind CtxKind) {
-  if (Res.Declaration)
+  if (Res.Declaration) {
+    if (extractFunctorCallOperator(Res.Declaration))
+      return CompletionItemKind::Function;
     return toCompletionItemKind(index::getSymbolInfo(Res.Declaration).Kind);
+  }
   if (CtxKind == CodeCompletionContext::CCC_IncludedFile)
     return CompletionItemKind::File;
   switch (Res.Kind) {
```

Passes all tests except CodeCompleteTests.cpp > L. 3480 > TEST(CompletionTest, Lambda) that I modified to expect a CompletionItemKind::Function.
Fixes my issue in VsCode when using this build locally.


Another idea would be to add a CompletionItemKind specific to Lambdas / Functors but it would modify behavior in places where lambas are currently expected as CompletionItemKind::Variable.

**Comment by HighCommander4** [comment]
> What do you think about fixing the problem before summarizeSnippet by detecting lambdas as CompletionItemKind::Function ?

The main impact I can think of that this would have, is that editors may display a different icon (function rather than variable) for completion items which are lambdas.

Not a huge deal, but I wonder if having different icons for function and lambdas may actually be a useful signal for some users (whereas lambdas can be distinguished from other variables already by having a signature)?

Returning to the original idea:

> The idea above fails some tests with template variable completion.

Perhaps an easy way to address this could be to amend [this condition](https://searchfox.org/llvm/rev/36cb9894b3eec301c1b2f92b5c05519ed1cea4fa/clang-tools-extra/clangd/CodeComplete.cpp#631) to something like `if (MayHaveArgList && llvm::StringRef(*Snippet)->contains("("))`?

That should avoid variable templates taking that branch.

**Comment by Nodli** [comment]
This does work and passes all supported tests on my machine ; fixes the issue in vscode.
I created a pull request https://github.com/llvm/llvm-project/pull/181009

Thank you for the support and for your input 🙂
